### PR TITLE
[receiver/vcenter] Adds vCenter VSAN Host metrics

### PR DESCRIPTION
**Pull request by BominRahmani**
**Description:** <Describe what has changed.>

The following PR adds VSAN host metric collection. The following metrics are collected 
```
vcenter.host.vsan.throughput          (direction={read/write})
vcenter.host.vsan.iops                (direction={read/write})
vcenter.host.vsan.congestions
vcenter.host.vsan.cache.hit_rate
vcenter.host.vsan.latency.avg         (direction={read/write})
```

**Link to tracking Issue:** #33556

**Testing:** <Describe what testing was performed and which tests were added.>
Added mock data fields in order to test vsan host metric scraping.  Updated golden test file output.

**Documentation:** <Describe the documentation added.>
Updated documentation using `make generate`

### CODE REVIEW

File: receiver/vcenterreceiver/client.go [hunk 1]
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
...
+	}
+
+	for _, clusterRef := range clusterRefs {
+		results, err := vc.vsanQueryByCluster(ctx, queryType, clusterRef)
+		if err != nil {
+			vsanErr := vc.handleVSANError(err, queryType)
+			if vsanErr != nil {
+				return nil, vsanErr
+			}
+			continue
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/client.go:413]: This continue is probably fine (compared to returning "nil, &allResults"). My only concern here is that we don't have a test for this situation right now. To be fair, we also don't have a test for the happy path of multiple clusters either for vSAN as well. It would be a bit of a PITA, but it might be worth modifying the scrape tests/results to have a 2nd cluster with a single resource pool/host/VM.

Really....this would normally be handled in the client tests with the simulator, but the vSAN simulator isn't working.

File: receiver/vcenterreceiver/client.go [hunk 2]
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
...
+		}
+
+		maps.Copy(allResults.MetricResultsByUUID, results.MetricResultsByUUID)
+	}
+
+	return &allResults, nil
+}
+
+// vsanQueryByCluster performs a VSAN query for a specific cluster
+func (vc *vcenterClient) vsanQueryByCluster(
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/client.go:423]: I can guess why you took the vSANDriver nil check out of this function, but this function should be able to stand on its own. So I would put it back here.

File: receiver/vcenterreceiver/client.go [hunk 3]
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
...
+			EntityRefId: string(queryType),
+			StartTime:   &now,
+			EndTime:     &now,
+			Labels:      vc.getLabelsForQueryType(queryType),
+		},
+	}
+
+	rawResults, err := vc.vsanDriver.VsanPerfQueryPerf(ctx, clusterRef, querySpec)
+	if err != nil {
+		return nil, fmt.Errorf("problem retrieving %s VSAN metrics for cluster %s: %w", queryType, clusterRef.Value, err)
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/client.go:444]: vSAN

File: receiver/vcenterreceiver/client.go [hunk 4]
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
...
+	}
+
+	for _, rawResult := range rawResults {
+		metricResults, err := vc.convertVSANResultToMetricResults(rawResult)
+		if err != nil {
+			return &queryResults, fmt.Errorf("problem processing %s VSAN metrics for cluster %s: %w", queryType, clusterRef.Value, err)
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/client.go:450]: vSAN

Comment (StefanKurek) [reply receiver/vcenterreceiver/client.go:450]: Also, this does simplify things from also checking for `metricResults == nil`, but you lose the ability to check for a specific problem UUID

File: receiver/vcenterreceiver/client.go [hunk 5]
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
...
+		}
+	default:
+		return []string{}
+	}
+}
+
+// VSANQuery performs a VSAN query for the specified type across multiple clusters
+func (vc *vcenterClient) VSANQuery(
+	ctx context.Context,
+	queryType VSANQueryType,
```
[the remark this answers is not in the document]
Comment (StefanKurek) [reply receiver/vcenterreceiver/client.go:395]: Ok as I keep reviewing, I see that you will have individual functions calling this one. So you could make this one private if it's never intended to be called outside of the Client `vSANQuery`

File: receiver/vcenterreceiver/scraper.go [hunk 6]
```diff
@@ -262,6 +266,22 @@ func (v *vcenterMetricScraper) scrapeHosts(ctx context.Context, dc *mo.Datacente
 		errs.AddPartial(1, fmt.Errorf("failed to retrieve perf metrics for HostSystems: %w", err))
 		return
 	}
+
+	// Get all host vSAN metrics and store for later retrieval
+	clusterRefs := []*types.ManagedObjectReference{}
+	for _, compute := range v.scrapeData.computesByRef {
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/scraper.go:272]: We could make a little helper function for this so it's not duplicated code.

File: receiver/vcenterreceiver/scraper.go [hunk 7]
```diff
@@ -262,6 +266,22 @@ func (v *vcenterMetricScraper) scrapeHosts(ctx context.Context, dc *mo.Datacente
 		errs.AddPartial(1, fmt.Errorf("failed to retrieve perf metrics for HostSystems: %w", err))
 		return
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/scraper.go:267]: I don't think you want to return here. Look at the VM function

File: receiver/vcenterreceiver/metadata.yaml [hunk 8]
```diff
@@ -421,6 +443,51 @@ metrics:
...
+    description: The vSAN IOPs of a host.
+    unit: "{operations/sec}"
+    gauge:
+      value_type: int
+    attributes: [vsan_operation_type]
+    warnings:
+      if_enabled_not_set: "this metric will be enabled by default starting in release v0.106.0"
+  vcenter.host.vsan.latency.avg:
+    enabled: false
+    description: The host latency while accessing VSAN storage.
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/metadata.yaml:466]: vSAN

File: receiver/vcenterreceiver/metadata.yaml [hunk 9]
```diff
@@ -421,6 +443,51 @@ metrics:
...
+    unit: "us"
+    gauge:
+      value_type: int
+    attributes: [vsan_latency_type]
+    warnings:
+      if_enabled_not_set: "this metric will be enabled by default starting in release v0.106.0"
+  vcenter.host.vsan.congestion:
+    enabled: false
+    description: The congestions of IOs generated by all vSAN clients in the host.
+    unit: "{congestions}"
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/metadata.yaml:476]: because this isn't given as a counter i think we will want to convert this to `congestions/sec`

File: receiver/vcenterreceiver/metrics.go [hunk 10]
```diff
@@ -406,3 +406,57 @@ func (v *vcenterMetricScraper) recordVMPerformanceMetrics(entityMetric *performa
...
+	for _, metric := range vSANMetrics.MetricDetails {
+		for i, value := range metric.Values {
+			timestamp := metric.Timestamps[i]
+			switch metric.MetricLabel {
+			case "iopsRead":
+				v.mb.RecordVcenterHostVsanOperationsDataPoint(pcommon.NewTimestampFromTime(*timestamp), value, metadata.AttributeVsanOperationTypeRead)
+			case "iopsWrite":
+				v.mb.RecordVcenterHostVsanOperationsDataPoint(pcommon.NewTimestampFromTime(*timestamp), value, metadata.AttributeVsanOperationTypeWrite)
+			case "throughputRead":
+				readRate := float64(value) / 20
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/metrics.go:446]: we're now pulling the metric interval so you'll want to divide by that just like in the VMs

File: receiver/vcenterreceiver/metrics.go [hunk 11]
```diff
@@ -406,3 +406,57 @@ func (v *vcenterMetricScraper) recordVMPerformanceMetrics(entityMetric *performa
...
+				v.mb.RecordVcenterHostVsanThroughputDataPoint(pcommon.NewTimestampFromTime(*timestamp), readRate, metadata.AttributeVsanThroughputDirectionRead)
+			case "throughputWrite":
+				writeRate := float64(value) / 20
+				v.mb.RecordVcenterHostVsanThroughputDataPoint(pcommon.NewTimestampFromTime(*timestamp), writeRate, metadata.AttributeVsanThroughputDirectionWrite)
+			case "latencyRead":
+				v.mb.RecordVcenterHostVsanLatencyAvgDataPoint(pcommon.NewTimestampFromTime(*timestamp), value, metadata.AttributeVsanLatencyTypeRead)
+			case "latencyWrite":
+				v.mb.RecordVcenterHostVsanLatencyAvgDataPoint(pcommon.NewTimestampFromTime(*timestamp), value, metadata.AttributeVsanLatencyTypeWrite)
+			case "congestion":
+				v.mb.RecordVcenterHostVsanCongestionDataPoint(pcommon.NewTimestampFromTime(*timestamp), value)
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/metrics.go:456]: Also divide this by interval I think

**Review by StefanKurek — CHANGES_REQUESTED** [review]

File: receiver/vcenterreceiver/client.go [hunk 12]
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
...
+// VSANQueryType represents the type of VSAN query
+type VSANQueryType string
+
+const (
+	VSANQueryTypeHosts           VSANQueryType = "host-domclient:*"
+	VSANQueryTypeVirtualMachines VSANQueryType = "virtual-machine:*"
+)
+
+// getLabelsForQueryType returns the appropriate labels for each query type
+func (vc *vcenterClient) getLabelsForQueryType(queryType VSANQueryType) []string {
```
[the remark this answers is not in the document]
Comment (StefanKurek) [reply receiver/vcenterreceiver/client.go:375]: Ok resolving based on what I saw later on.

File: receiver/vcenterreceiver/client.go [hunk 13]
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
...
+	return vc.VSANQuery(ctx, VSANQueryTypeVirtualMachines, clusterRefs)
+}
+
+func (vc *vcenterClient) handleVSANError(
+	err error,
+	queryType VSANQueryType,
+) error {
+	faultErr := errors.Unwrap(err)
+	if faultErr == nil {
+		return fmt.Errorf("Problem retrieving %s vSan metrics: %w", queryType, err)
```
[the remark this answers is not in the document]
Comment (BominRahmani) [reply receiver/vcenterreceiver/client.go:481]: Good to know!

File: receiver/vcenterreceiver/client.go (hunk 1 above)
Comment (BominRahmani) [reply receiver/vcenterreceiver/client.go:413]: I can modify the scrape tests, yea its a shame this won't be handled with the govmomi simulator, but I think this is a close second best alternative.

File: receiver/vcenterreceiver/client.go (hunk 4 above)
Comment (BominRahmani) [reply receiver/vcenterreceiver/client.go:450]: Do you think the simplicity warrants the lack of extra information?

File: receiver/vcenterreceiver/metadata.yaml (hunk 9 above)
Comment (BominRahmani) [reply receiver/vcenterreceiver/metadata.yaml:476]: Can you explain this to me? I thought the congestions were a counter?

File: receiver/vcenterreceiver/client.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (StefanKurek) [reply receiver/vcenterreceiver/client.go:395]: resolve

File: receiver/vcenterreceiver/client.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (StefanKurek) [reply receiver/vcenterreceiver/client.go]: resolve

File: receiver/vcenterreceiver/client.go (hunk 1 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/client.go:413]: @BominRahmani Soooo, maybe you should take a look at the VM/Cluster PRs that I have out. I no longer am passing in a list of clusters, and this continue isn't actually needed...

File: receiver/vcenterreceiver/client.go (hunk 2 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/client.go:423]: Again, in my other PRs this function doesn't even exist...

File: receiver/vcenterreceiver/client.go (hunk 4 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/client.go]: Extra information when a customer has a problem is often great to have.

File: receiver/vcenterreceiver/scraper.go (hunk 6 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/scraper.go]: In my new PRs this doesn't exist as we don't have to worry about passing in cluster anymore

File: receiver/vcenterreceiver/scraper.go [hunk 14]
```diff
@@ -264,17 +264,12 @@ func (v *vcenterMetricScraper) scrapeHosts(ctx context.Context, dc *mo.Datacente
 	results, err := v.client.PerfMetricsQuery(ctx, spec, hostPerfMetricList, hsRefs)
 	if err != nil {
 		errs.AddPartial(1, fmt.Errorf("failed to retrieve perf metrics for HostSystems: %w", err))
-		return
+	} else {
+		v.scrapeData.hostPerfMetricsByRef = results.resultsByRef
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/scraper.go:268]: Because you added this line here, you want to remove the one below

File: receiver/vcenterreceiver/metadata.yaml (hunk 9 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/metadata.yaml:476]: Let me rephrase. Because it isn't given as a sum counter (from the start of time) vs a delta counter (over a somewhat arbitrary interval).

File: receiver/vcenterreceiver/metrics.go (hunk 10 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/metrics.go]: resolve

File: receiver/vcenterreceiver/client.go [hunk 15]
```diff
@@ -464,17 +436,15 @@ func (vc *vcenterClient) vsanQueryByCluster(
 // VSANHosts returns back host VSAN performance metrics for a group of clusters
 func (vc *vcenterClient) VSANHosts(
 	ctx context.Context,
-	clusterRefs []*vt.ManagedObjectReference,
 ) (*VSANQueryResults, error) {
-	return vc.vSANQuery(ctx, VSANQueryTypeHosts, clusterRefs)
+	return vc.vSANQuery(ctx, VSANQueryTypeHosts)
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/client.go:440]: @BominRahmani unless you use the handleVSANError func inside here, you'll never turn any of the "lesser" errors to logs (same for the VM func)

File: receiver/vcenterreceiver/metrics.go [hunk 16]
```diff
@@ -453,7 +453,8 @@ func (v *vcenterMetricScraper) recordHostVSANMetrics(vSANMetrics *VSANMetricResu
 			case "latencyWrite":
 				v.mb.RecordVcenterHostVsanLatencyAvgDataPoint(pcommon.NewTimestampFromTime(*timestamp), value, metadata.AttributeVsanLatencyTypeWrite)
 			case "congestion":
-				v.mb.RecordVcenterHostVsanCongestionDataPoint(pcommon.NewTimestampFromTime(*timestamp), value)
+				congestionRate := float64(value) / float64(metric.Interval)
+				v.mb.RecordVcenterHostVsanCongestionDataPoint(pcommon.NewTimestampFromTime(*timestamp), int64(congestionRate))
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/metrics.go:457]: Shouldn't be converting congestionRate to an int64. it should be a float in the metadata.yaml

File: receiver/vcenterreceiver/scraper.go (hunk 14 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/scraper.go:268]: @BominRahmani you still need to remove the line on 277

File: receiver/vcenterreceiver/testdata/metrics/expected-all-enabled.yaml [hunk 17]
```diff
@@ -2924,14 +2924,14 @@ resourceMetrics:
           - description: The congestions of IOs generated by all vSAN clients in the host.
             gauge:
               dataPoints:
-                - asInt: "1"
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/testdata/metrics/expected-all-enabled.yaml]: these are only 0 because you're converting to int. you could pick larger values in your mocked responses so they nicely divide by the interval (300) to get whole number results here, but not necessary

**Review by StefanKurek — CHANGES_REQUESTED** [review]
Close. A few small changes left, but looking pretty good.

File: receiver/vcenterreceiver/client.go (hunk 1 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/client.go]: resolve

File: receiver/vcenterreceiver/scraper.go (hunk 6 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/scraper.go]: resolve

File: receiver/vcenterreceiver/metadata.yaml (hunk 9 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/metadata.yaml]: resolve

File: receiver/vcenterreceiver/client.go (hunk 15 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/client.go]: resolve

File: receiver/vcenterreceiver/scraper.go (hunk 14 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/scraper.go:268]: resolve

**Review by StefanKurek — APPROVED** [review]

File: receiver/vcenterreceiver/testdata/metrics/expected-all-enabled.yaml (hunk 17 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/testdata/metrics/expected-all-enabled.yaml]: resolve

File: receiver/vcenterreceiver/metrics.go (hunk 16 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/metrics.go]: resolve

**Review by djaglowski — APPROVED** [review]
